Project template (#1)
betrybe/sd-01-week12-movie-card-library-crud: README.md

Uma rota define o que deve ser renderizado na página ao abri-la. Cada rota está associada a um caminho. O caminho é a parte da URL após o domínio (nome do site, de forma simplificada). Por exemplo, em `www.example.com/foo/bar`, o caminho é `/foo/bar`. Até agora, todos os apps React que você desenvolveu possuiam somente uma rota, a raiz (`/`). 

Este app terá 4 rotas:

1. A rota raiz, no caminho `/`. Esta rota exibirá uma lista com todos os filmes cadastrados. Essa lista contém informações resumidas sobre cada filme.

2. Uma rota para mostrar informações detalhadas de um filme, no caminho `movies/:id`. `:id`, um parâmetro de URL, é o id do filme exibido. Por exemplo, ao entrar no caminho `/movies/5`, serão exibidas informações sobre o filme com id 5. 

3. Uma rota para criar novos filmes, no caminho `movies/new`. Essa rota renderizará um formulário para adicionar um novo filme.

> 4. Uma rota para editar um filme, no caminho `movies/:id/edit`. Analogamente à rota 2, `:id` é o id do filme a ser editado. Essa rota renderizará um formulário idêntico ao da rota 3. Nesse caso, porém, o formulário virá preenchido com as informações do filme a ser editado. Ao submeter o formuário, ao invés de criar um novo filme, o filme em questão terá seus dados atualziados.

Reviewer: ```suggestion
4. Uma rota para editar um filme, no caminho `movies/:id/edit`. Analogamente à rota 2, `:id` é o id do filme a ser editado. Essa rota renderizará um formulário idêntico ao da rota 3. Nesse caso, porém, o formulário virá preenchido com as informações do filme a ser editado. Ao submeter o formulário, ao invés de criar um novo filme, o filme em questão terá seus dados atualizados.
```

> Relacionado a cada rota haverá um componente React responsável por rendezirar seu conteúdo. Esse mapeamento entre o caminho da URL, rota e componente será feito pelo `React Router`, a principal biblioteca de roteamento em `React`.

Reviewer 2: ```suggestion
Relacionado a cada rota haverá um componente React responsável por renderizar seu conteúdo. Esse mapeamento entre o caminho da URL, rota e componente será feito pelo `React Router`, a principal biblioteca de roteamento em `React`.
```

> Naturalmente, haverá links de navegação e redirecionamento entre as diferentes rotas. Por exemplo, na rota 1, haverá, para cada filme, um link para a rota 2, onde se poderá ver informações detalhadas sobre o filme escolhido. Na rota 2, haverá um link para a rota 4, a fim de se editar informações do filme. Ao submeter o formulário, o app automaticamente será levado de volta a rota 2, mostrando as informações atualizadas do filme. Tudo isso será feito utilizando os componenetes da biblioteca `React Rouer`.

Reviewer 2: ```suggestion
Naturalmente, haverá links de navegação e redirecionamento entre as diferentes rotas. Por exemplo, na rota 1, haverá, para cada filme, um link para a rota 2, onde se poderá ver informações detalhadas sobre o filme escolhido. Na rota 2, haverá um link para a rota 4, a fim de se editar informações do filme. Ao submeter o formulário, o app automaticamente será levado de volta à rota 2, mostrando as informações atualizadas do filme. Tudo isso será feito utilizando os componentes da biblioteca `React Router`.
```

## Requisitos do projeto

### 1 O componente `App` deve renderizar `BrowserRouter`

O componente `App`, que representa toda a aplicação, deve importar e renderizar `BrowserRouter`. 

### 2 O componente `BrowserRouter` deve renderizar `Switch`

O componente `Switch` deve ser o único filho direto de `BrowserRouter`. Dentro de `Switch` ficarão as rotas relacionadas às diferentes páginas da aplicação.

### 3 O componente `Switch` deve renderizar as 4 rotas do app

O app possui ao todo 4 rotas, como descrito na seção [O que deverá ser desenvolvido](#o-que-deverá-ser-desenvolvido). Cada rota é associada a um caminho de URL e a um componente. Este componente renderizará um conteúdo específico para aquela rota. O mapeamento entre rotas, caminhos de URL e componentes estão listados abaixo.

1. A caminho raiz do projeto (`/`) deve renderizar o componente `MovieList`.

> 2. O caminho `movies/:movieId` deve renderizar o componentes `MovieDetails`. `:id` é um parâmtro de URL com o id do filme que `MovieDetails` renderizará.

Reviewer: ```suggestion
2. O caminho `movies/:movieId` deve renderizar o componente `MovieDetails`. `:id` é um parâmetro de URL com o id do filme que `MovieDetails` renderizará.
```

> 3. O caminho `movies/new` deve renderiar o componente `NewMovie`.

Reviewer 2: ```suggestion
3. O caminho `movies/new` deve renderizar o componente `NewMovie`.
```

Author: Already corrected from Nobre's suggestion. 👍

> 4. O caminho `movies/:movieId/edit` devve renderizar o componente `EditMovie`. `:id` é um parâmetro de URL com o id do filme que `EditMovie` possibibilitará editar.

Reviewer: ```suggestion
4. O caminho `movies/:movieId/edit` deve renderizar o componente `EditMovie`. `:id` é um parâmetro de URL com o id do filme que `EditMovie` possibilitará editar.
```

5. Qualquer outro caminho que não se enquadre nas rotas anteriores deve renderizar o componente `NotFound`.

### 4 Ao ser montado, `MovieList` deve fazer uma requisição para buscar a lista de filmes a ser renderizada

Para buscar a lista, você deve utilizar a função `getMovies` importada do módulo `movieAPI` em `MovieList`. Essa função returna uma _promise_. A requisição deve ser feita no momento em que o `MovieList` for montado no DOM. Enquanto a requisição estiver em curso, `MovieList` deve renderizar o componente `Loading`, como ilustrado na imagem a seguir.

![image](loading.png)

Uma vez que a requisição retornar, o resultado deve ser renderizado. Para cada filme, renderize um componente `MovieCard`, como ilustrado abaixo.

![image](index.png)

Você precisará adicionar estado em `MovieList` para controlar o que será exibido.

### 5 `MovieCard` deve possuir um link para a página de detalhes de um filme

> Cada cartão da lista deve conter em seu rodapé um link com o texto "VER DETALHES". Este link deve apontar para a rota `movies/:id`, onde `:id` é o id do filme. Esta rota exibirá informáções detalhadas de um filme.

Reviewer: ```suggestion
Cada cartão da lista deve conter em seu rodapé um link com o texto "VER DETALHES". Este link deve apontar para a rota `movies/:id`, onde `:id` é o id do filme. Esta rota exibirá informações detalhadas de um filme.
```

### 6 Ao ser montado, `MovieDetails` deve fazer uma requisição para buscar o filme que deverá ser renderizado

> `MovieDetails` se comporta de forma muito semelhante a `MovieList`. Ao ser montado, deve fazer uma requisição utilizando a função `getMovie` do módulo `movieAPI`, passando o id do filme. O componente `Loading` deve ser renderizado enquando a requisição estiver em curso. Após terminar, deve-se renderizar o card com detalhes sobre o filme.

Reviewer: ```suggestion
`MovieDetails` se comporta de forma muito semelhante a `MovieList`. Ao ser montado, deve fazer uma requisição utilizando a função `getMovie` do módulo `movieAPI`, passando o id do filme. O componente `Loading` deve ser renderizado enquanto a requisição estiver em curso. Após terminar, deve-se renderizar o card com detalhes sobre o filme.
```

### 7 `MovieDetails` deve possuir um link a página de edição de filme.

> No rodapé do cartão, deve haver um link com o texto "EDITAR" apontando para a rota `/movies/:movieId/edit`, conforme ilustrado na imagem abaixo.

Author: The parameter is in fact called `movieId` throughout the code and README

### 10 Ao submeter o formulário, `EditMovie` deve atualizar o filme e redirecionar a página para a rota raiz.

O componente `MovieForm` recebe uma callback, que será executada quando o formulário for submetido. Essa callback recebe os dados atualizados do filme. Neste momento, `EditMovie` deve fazer uma requisição utilizando a função `updateMovie` do módulo `movieAPI`, passando como argumentos os dados atualizados. Quando a requisição finalizar, `EditMovie` deve redirecionar o app para a rota raiz (`/`).

### 11 Na página inicial, deve haver um link para criar novos cartões.

O link deve conter o texto "ADICIONAR CARTÃO" e apontar para a rota `movies/new`, contento um formulário para criar novos cartões.

### 12 Ao submeter o formulário, `NewMovie` deve criar um novo filme e redirecionar o app para a página inicial.

Utilizando a callback passada para `MovieForm`, `NewMovie` deve criar um novo cartão utilizando a função `createMovie` do módulo `movieAPI`. Após o fim da requisição, `NewMovie` deve redirecionar o app para a página inicial, contento o novo cartão.

### 13 Adicione proptypes a todos os componentes

> Todos os compontens que recebem props devem ter suas proptypes corretamente declaradas. O _eslint_ checa automaticamete declaração de proptypes, portanto seu _Pull Request_ deverá passar no _Code Climate para satisfazer esse requisito.

Reviewer 3: Sorry, I misspeled the word `automaticamente` again :joy:

Here is the correction:

```suggestion
Todos os compontens que recebem _props_ devem ter suas _proptypes_ corretamente declaradas. O _eslint_ checa automaticamente declaração de _proptypes_, portanto seu _Pull Request_ deverá passar no _Code Climate_ para satisfazer esse requisito.
```

    * Você deve criar uma branch no seguinte formato: `nome-de-usuario-nome-do-projeto`
    * Exemplo: `git checkout -b joaozinho-movie-card-library-crud`

4. TODO: Add initial step before uploading project on trybe organization

5. Adicione as mudanças ao _stage_ do Git e faça um `commit`
  * Verifique que as mudanças ainda não estão no _stage_
    * Exemplo: `git status` (deve aparecer listado o arquivo _src/components/MovieList.jsx_ em vermelho)
  * Adicione o arquivo alterado ao _stage_ do Git
      * Exemplo:
        * `git add .` (adicionando todas as mudanças - _que estavam em vermelho_ - ao stage do Git)
        * `git status` (deve aparecer listado o arquivo _src/components/MovieLibrary.jsx_ em verde)
  * Faça o `commit` inicial
      * Exemplo:
        * `git commit -m 'iniciando o projeto. VAMOS COM TUDO :rocket:'` (fazendo o primeiro commit)
        * `git status` (deve aparecer uma mensagem tipo _nothing to commit_ )

6. Adicione a sua branch com o novo `commit` ao repositório remoto
  * Usando o exemplo anterior: `git push -u origin joaozinho-movie-cards-library-crud`

7. Crie um novo `Pull Request` _(PR)_

>   * Vá até a página de _Pull Requests_ do [repositório no GitHub](https://github.com/tryber/sd-01-week11-movie-card-library-crud/pulls)

Reviewer 2: ```suggestion
  * Vá até a página de _Pull Requests_ do [repositório no GitHub](https://github.com/tryber/sd-01-week12-movie-card-library-crud/pulls)
```

  * Clique no botão verde _"New pull request"_
  * Clique na caixa de seleção _"Compare"_ e escolha a sua branch **com atenção**
  * Clique no botão verde _"Create pull request"_
  * Adicione uma descrição para o _Pull Request_ e clique no botão verde _"Create pull request"_
  * **Não se preocupe em preencher mais nada por enquanto!**

>   * Volte até a [página de _Pull Requests_ do repositório](https://github.com/tryber/sd-01-week11-movie-card-library-crud/pulls) e confira que o seu _Pull Request_ está criado

Reviewer 2: ```suggestion
  * Volte até a [página de _Pull Requests_ do repositório](https://github.com/tryber/sd-01-week12-movie-card-library-crud/pulls) e confira que o seu _Pull Request_ está criado
```
